src/cpp/src/sampler.cpp

```diff
@@ -587,45 +590,95 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
+                            break;
+                        } 
+                        Token sampled_token_id;
+                        // get logit processors only for validation or generation, not for cases of extending KV cache
+                        if (token_id_per_seq == 0 && !is_validation_mode_enabled || is_validation_mode_enabled) {
+                            auto logit_vector = _get_logit_vector(sequence_group_logits, running_sequence_id, token_id_per_seq);
+                            logit_processor.apply(logit_vector);
+                            Token sampled_token_id;
+                            if (sampling_params.is_greedy_decoding()) {
+                                sampled_token_id = _greedy_sample(logit_vector);
+                            } else {
+                                // is_multinomial()
+                                const bool is_generate_n_tokens = sequence_group->num_total_seqs() == 1;
+                                const size_t num_tokens_per_sequence = is_generate_n_tokens ? sampling_params.num_return_sequences : 1;
+                                auto sampled_token_ids = _multinomial_sample(logit_vector, num_tokens_per_sequence);
+                                sampled_token_id = sampled_token_ids[0];
+
+                                if (is_generate_n_tokens) {
+                                    auto sequence_to_fork = running_sequences[0];
+                                    std::list<uint64_t> forked_seq_ids;
+                                    for (size_t i = num_running_sequences; i < num_tokens_per_sequence; ++i) {
+                                        const auto forked_sequence = sequence_group->fork_sequence(sequence_to_fork);
+                                        const auto forked_seq_id = forked_sequence->get_id();
+                                        forked_seq_ids.push_back(forked_seq_id);
+                                        running_sequences = sequence_group->get_running_sequences();
+                                        register_new_token(sampled_token_ids[i], forked_sequence->get_grouped_id(), true);
+                                    }
+                                    sampler_output.m_forked_sequences.insert({running_sequences[0]->get_id(), forked_seq_ids});
+                                }
+                            }
+                            // flag to add sampled token to generated sequence or extend logit processors only
+                            bool is_extend_sequence = token_id_per_seq == 0;
+                            if (token_id_per_seq > 0) {
+                                auto it = running_seq_token_ids.rbegin();
+                                std::advance(it, token_id_per_seq - 1);
+                                // to validate candidates from assisting model and remove incorrect ones from generated sequence
+                                if (is_validation_mode_enabled && *it != sampled_token_id.m_index) {
```

Comment:
in general, the sampler logic become even more complex as now it's responsible for validation (part of speculative decoding).

Is it possible to implement speculative part on top of existing sampler functionality (which is just to sampler new tokens w/o any verification, removing tokens from end of sequences) ?

Maybe we need to split implementation of current sampler into parts:
- provide next token based on logits tensor - basic building block, which does not depend on sequence group. I suppose this building block can be also reused in LLM stateful pipeline, Visual language and Whisper pipelines.
- then, adding this new sampled token to sequence group. This bullet implies that output of sampler is consumed by sequence group as is.

In this case, speculative sampler logic can be implemented on top of building blocks from bullet 1 and "basic" sampler (bullet 2) will not be "overcompilcated" by speculative logic

Reply:
I have separated some functionality from `sample` method to stand alone functions defined in the cpp file. The changes allows to simplify the method and improve readability. Please check